utils/exception_container.hh

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include <exception>
+#include <type_traits>
+#include <memory>
+#include <variant>
+#include <seastar/core/future.hh>
+#include <seastar/core/distributed.hh>
+#include "utils/variant_element.hh"
+
+namespace utils {
+
+class bad_exception_container_access : public std::exception {
+public:
+    const char* what() const noexcept override {
+        return "bad exception container access";
+    }
+};
+
+// A variant-like type capable of holding one of the allowed exception types.
+// This allows inspecting the exception in the error handling code without
+// having to resort to costly rethrowing of std::exception_ptr, as is
+// in the case of the usual exception handling.
+//
+// It's not as ergonomic as using exceptions with seastar, but allows for
+// fast inspection and manipulation.
+//
+// Exceptions are held in a std::variant. The variant is kept behind
+// a std::unique_ptr to ensure that moves are cheap. This means that
+// a moved-out exception container becomes "empty" and does not contain
+// a valid exception.
```

Comment:
I thought about your suggestion a bit more and I think it still is worth considering. While I like that exception_container explicitly lists possible exceptions, just keeping a shared_ptr\<std::exception\> or something in that fashion may be more consistent with how real C++ exceptions work.

That said, I'd rather try doing it in a followup. I don't think it will be too much work to convert the current approach to it, and the performance impact of the shared_ptr\<std::exception\> approach needs to be evaluated.